### PR TITLE
Add daily/weekly/monthly leaderboards

### PR DESCRIPTION
**Pull request by taahamahdi**
Also adds pagination to leaderboard

### CODE REVIEW

File: src/helpers/management_utils.ts [hunk 1]
```diff
@@ -257,6 +259,17 @@ export async function clearRestartNotification() {
         .update({ restart_time: null });
 }
 
+/**
+ * Clears the given temporary scoreboard on cronjob timer
+ * Sends scoreboard before deletion in debug channel
+ */
+async function clearTemporaryLeaderboard(duration: LeaderboardDuration.DAILY | LeaderboardDuration.WEEKLY | LeaderboardDuration.MONTHLY) {
+    await new LeaderboardCommand().showLeaderboard(
```
Comment (Brainicism) [inline src/helpers/management_utils.ts:267]: Why isn't `showLeaderboard` static?

File: src/commands/game_commands/leaderboard.ts [hunk 2]
```diff
@@ -2,30 +2,42 @@ import Eris from "eris";
...
     ENROLL = "enroll",
-    UNEROLL = "unenroll",
-    PAGE = "page",
-    SERVER = "server",
-    GAME = "game",
+    UNENROLL = "unenroll",
 }
 
+export const TABLE_BY_DURATION = {
+    [LeaderboardDuration.INDEFINITE]: "player_stats",
```
Comment (Brainicism) [inline src/commands/game_commands/leaderboard.ts:35]: permanent?

File: src/migrations/20210828065453_daily_player_stats_table.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
+exports.up = function (knex) {
+    return knex.schema.createTableIfNotExists("daily_player_stats", function (table) {
+        table.string("player_id").notNullable();
+        table.integer("songs_guessed").defaultTo(0);
+        table.integer("games_played").defaultTo(0);
+        table.integer("exp").defaultTo(0);
+        table.integer("level").defaultTo(0);
```
Comment (Brainicism) [inline src/migrations/20210828065453_daily_player_stats_table.js:7]: What's the purpose of recording level gains? Doesn't scale linearly

File: src/structures/game_session.ts [hunk 4]
```diff
@@ -960,6 +960,21 @@ export default class GameSession {
             .onConflict("player_id")
             .ignore();
 
+        await dbContext.kmq("daily_player_stats")
```
Comment (Brainicism) [inline src/structures/game_session.ts:963]: Instead of having 3 different tables, can't we use the same table, and filter daily/weekly/monthly with a WHERE condition on date?

File: src/commands/game_commands/leaderboard.ts [hunk 5]
```diff
@@ -35,32 +47,32 @@ export default class LeaderboardCommand implements BaseCommand {
...
             },
-            {
-                example: "`,leaderboard game`",
-                explanation: "Shows the leaderboard of players with points in the current game",
-            },
             {
                 example: "`,leaderboard game 2`",
                 explanation: "Shows the 2nd page of the players with points in the current game",
             },
+            {
```
Comment (Brainicism) [inline src/commands/game_commands/leaderboard.ts:72]: Inconsistency on where the page parameter goes is not user friendly
`,leaderboard 3 server`
`,leaderboard weekly 4`

**Review by Brainicism — CHANGES_REQUESTED** [review]

File: src/migrations/20210828065453_daily_player_stats_table.js (hunk 3 above)
Comment (taahamahdi) [reply src/migrations/20210828065453_daily_player_stats_table.js:7]: It displays the levels gained. It is still a nice to have secondary stat in the leaderboard

File: src/commands/game_commands/leaderboard.ts (hunk 5 above)
Comment (taahamahdi) [reply src/commands/game_commands/leaderboard.ts:72]: I mainly added it to show order doesn't matter, would you prefer having the page number at the end for both of them?

File: src/migrations/20210828065453_temporary_player_stats_table.js [hunk 6]
```diff
@@ -0,0 +1,14 @@
+exports.up = function (knex) {
+    return knex.schema.createTableIfNotExists("temporary_player_stats", function (table) {
```
Comment (Brainicism) [inline src/migrations/20210828065453_temporary_player_stats_table.js:2]: player_game_session_stats

File: src/structures/game_session.ts [hunk 7]
```diff
@@ -947,40 +947,65 @@ export default class GameSession {
...
+                .transacting(trx);
 
-        await dbContext.kmq("player_servers")
-            .insert({
-                player_id: userID,
-                server_id: this.guildID,
-            })
-            .onConflict(["player_id", "server_id"])
-            .ignore();
+            await dbContext.kmq("temporary_player_stats")
```
Comment (Brainicism) [inline src/structures/game_session.ts:965]: Don't need to initialize it. The `currentDateString` here and in `incrementPlayerSongsGuessed` and `incrementPlayerExp` may be different. Just insert an entry once per game session end per player

**Review by Brainicism — CHANGES_REQUESTED** [review]

File: src/structures/game_session.ts (hunk 7 above)
Comment (taahamahdi) [reply src/structures/game_session.ts:965]: Can I create a date object and pass it to the functions that increment song guesses, exp gain, and the initialization? It maintains good abstraction in the for loop of endSession compared to a single insert

File: src/commands/game_commands/leaderboard.ts [hunk 8]
```diff
@@ -35,14 +41,14 @@ export default class LeaderboardCommand implements BaseCommand {
     help = {
         name: "leaderboard",
         description: "View the KMQ leaderboard.",
-        usage: ",leaderboard\n,leaderboard [server | game] {page_number}\n,leaderboard page {page_number}\n,leaderboard [enroll | unenroll]",
+        usage: ",leaderboard {server | game} {page_number} {daily | weekly | monthly}\n,leaderboard [enroll | unenroll]",
```
Comment (Brainicism) [inline src/commands/game_commands/leaderboard.ts:44]: This usage is very confusing.. maybe split into another line 
`,leaderboard {page_number}`
`,leaderboard {server | game | daily | weekly | monthly} {page_number}`
`,leaderboard [enroll | uneroll]`

File: src/commands/game_commands/leaderboard.ts [hunk 9]
```diff
@@ -2,28 +2,34 @@ import Eris from "eris";
...
 
-enum LeaderboardType {
+export enum LeaderboardType {
     GLOBAL = "global",
     SERVER = "server",
     GAME = "game",
 }
 
+export enum LeaderboardDuration {
+    PERMANENT = "permanent",
```
Comment (Brainicism) [inline src/commands/game_commands/leaderboard.ts:24]: ALL_TIME?

File: src/commands/game_commands/leaderboard.ts [hunk 10]
```diff
@@ -75,60 +81,56 @@ export default class LeaderboardCommand implements BaseCommand {
...
-            } else if (action === LeaderboardAction.GAME) {
-                LeaderboardCommand.showLeaderboard(message, 0, LeaderboardType.GAME);
-            }
+        let type = LeaderboardType.GLOBAL;
+        let duration = LeaderboardDuration.PERMANENT;
+        let pageOffset = 0;
 
-            return;
-        }
+        for (const arg of parsedMessage.components) {
```
Comment (Brainicism) [inline src/commands/game_commands/leaderboard.ts:100]: what in tarnation is this

File: src/commands/game_commands/leaderboard.ts [hunk 11]
```diff
@@ -154,83 +156,192 @@ export default class LeaderboardCommand implements BaseCommand {
...
-        }));
-
-        let leaderboardType: string;
-        switch (type) {
-            case LeaderboardType.GLOBAL:
-                leaderboardType = "Global";
+        return embedsFns;
+    }
+
+    private static async getPageCount(messageContext: MessageContext, type: LeaderboardType, duration: LeaderboardDuration): Promise<number> {
```
Comment (Brainicism) [inline src/commands/game_commands/leaderboard.ts:263]: why does getPageCount need to query again? in both cases it's been used, you have access to the result set

**Review by Brainicism — CHANGES_REQUESTED** [review]

File: src/commands/game_commands/leaderboard.ts (hunk 11 above)
Comment (taahamahdi) [reply src/commands/game_commands/leaderboard.ts:263]: I can remove one use of this function at:
`if (pageOffset + 1 > await LeaderboardCommand.getPageCount(messageContext, type, duration)) {`
Replacing it with the size of the array given by `getLeaderboardEmbeds`.

However, within `getLeaderboardEmbeds`, we don't have access to the result set (the query is awaited piece by piece in each anonymous function generated), so it still needs to be used there.

Comment (Brainicism) [reply src/commands/game_commands/leaderboard.ts:263]: Make `getLeaderboardEmbeds` return the size of `topPlayersQuery` too

**Review by Brainicism — CHANGES_REQUESTED** [review]

File: src/test/ci/leaderboard.test.ts [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+import KmqMember from "../../structures/kmq_member";
+
+describe("getLeaderboardEmbeds", () => {
+    describe("all-time leaderboard", () => {
+        afterEach(async () => {
+            await dbContext.kmq("player_stats").del();
+        });
+
+        it("should have one page of entries", async () => {
+            const entries = Math.floor((Math.random() * 50) / 10) * 10;
```
Comment (Brainicism) [inline src/test/ci/leaderboard.test.ts:16]: Tests shouldn't have randomness. If using randomness to account for edge case, the edge case should just be a different test

File: src/test/ci/leaderboard.test.ts [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+            for (let i = 0; i < entries; i++) {
+                await dbContext.kmq("player_stats")
```
Comment (Brainicism) [inline src/test/ci/leaderboard.test.ts:18]: Bulk insert instead of looping single inserts

File: src/test/ci/leaderboard.test.ts [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+import MessageContext from "../../structures/message_context";
+import KmqMember from "../../structures/kmq_member";
+
+describe("getLeaderboardEmbeds", () => {
+    describe("all-time leaderboard", () => {
+        afterEach(async () => {
+            await dbContext.kmq("player_stats").del();
+        });
+
+        it("should have one page of entries", async () => {
```
Comment (Brainicism) [inline src/test/ci/leaderboard.test.ts:15]: This test description seems inaccurate?

File: src/test/ci/leaderboard.test.ts [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+                        songs_guessed: i,
+                        exp: i,
+                        level: i,
+                    });
+            }
+
+            assert.strictEqual(Math.ceil(entries / 10), (await LeaderboardCommand.getLeaderboardEmbeds(new MessageContext("", new KmqMember("", "", "", ""), "", ""), LeaderboardType.GLOBAL, LeaderboardDuration.ALL_TIME)).pageCount);
+        });
+
+        it("should have two pages of entries", async () => {
```
Comment (Brainicism) [inline src/test/ci/leaderboard.test.ts:30]: This is the same test?

File: src/test/ci/leaderboard.test.ts [hunk 16]
```diff
@@ -0,0 +1,87 @@
+import assert from "assert";
+import { describe } from "mocha";
+import LeaderboardCommand, { LeaderboardType, LeaderboardDuration } from "../../commands/game_commands/leaderboard";
+import { getSqlDateString } from "../../helpers/discord_utils";
+import dbContext from "../../database_context";
+import MessageContext from "../../structures/message_context";
+import KmqMember from "../../structures/kmq_member";
+
+describe("getLeaderboardEmbeds", () => {
+    describe("all-time leaderboard", () => {
```
Comment (Brainicism) [inline src/test/ci/leaderboard.test.ts:10]: Also add test for multiple entries per player

File: src/test/ci/leaderboard.test.ts [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+                        exp: i,
+                        level: i,
+                    });
+            }
+
+            assert.strictEqual(Math.ceil(entries / 10), (await LeaderboardCommand.getLeaderboardEmbeds(new MessageContext("", new KmqMember("", "", "", ""), "", ""), LeaderboardType.GLOBAL, LeaderboardDuration.ALL_TIME)).pageCount);
+        });
+    });
+
+    describe("temporary leaderboard", () => {
```
Comment (Brainicism) [inline src/test/ci/leaderboard.test.ts:46]: tests seem identical? need tests that have different results based on dates

**Review by Brainicism — CHANGES_REQUESTED** [review]

File: src/test/ci/leaderboard.test.ts [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+            });
+        });
+
+        describe("server leaderboard", () => {
+            afterEach(async () => {
+                await dbContext.kmq("player_stats").del();
+                await dbContext.kmq("player_servers").del();
+            });
+
+            it("should match the number of pages", async () => {
```
Comment (Brainicism) [inline src/test/ci/leaderboard.test.ts:44]: Add some entries for a different server too

File: src/test/ci/leaderboard.test.ts (hunk 16 above)
Comment (Brainicism) [reply src/test/ci/leaderboard.test.ts:16]: Add coverage for potential off-by-one errors, `ENTRIES_PER_PAGE + 1`, `ENTRIES_PER_PAGE -1`

File: src/test/ci/leaderboard.test.ts [hunk 19]
```diff
@@ -0,0 +1,189 @@
...
+        describe("game leaderboard", () => {
+            afterEach(async () => {
+                await dbContext.kmq("player_stats").del();
+                await dbContext.kmq("player_servers").del();
+            });
+
+            it("should match the number of pages", async () => {
+                const gameSession = new GameSession("", "", SERVER_ID, gameStarter, GameType.CLASSIC);
+                state.gameSessions = { [SERVER_ID]: gameSession };
+                const entries = Math.floor((Math.random() * 50) / ENTRIES_PER_PAGE) * ENTRIES_PER_PAGE;
```
Comment (Brainicism) [inline src/test/ci/leaderboard.test.ts:75]: Include entries for people, not participating in the game

**Review by Brainicism — CHANGES_REQUESTED** [review]

**Review by Brainicism — APPROVED** [review]
